From d881d4f512780ef803543c79d10cbf6d71093d13 Mon Sep 17 00:00:00 2001
From: Sebastien PASDELOUP <sebastien.pasdeloup-ext@st.com>
Date: Mon, 22 Apr 2024 11:15:40 +0200
Subject: [PATCH 576/841] feat(st-crypto): add STM32MP21 SAES driver support
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

This patch makes this driver compatible with STM32MP21 SoC.
This implies bit-field modification and also support of 192 bit key
size. Please note that on STM32MP21, the CCF bit from ISR register is
no more replicated in SR register. Directly read CCF from ISR register.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/arm-trusted-firmware.git 9f6c1f02a2d45138c025b52f517e6e6a79e0b0ed feat(st-crypto): add STM32MP21 SAES driver support]
Signed-off-by: Maxime Méré <maxime.mere@foss.st.com>
Change-Id: Ic3a55c68e5052099b001e61adeca25edd887e631
Reviewed-on: https://gerrit.st.com/c/mpu/oe/st/tf-a/+/361800
Domain-Review: Sebastien PASDELOUP <sebastien.pasdeloup-ext@st.com>
Reviewed-by: Sebastien PASDELOUP <sebastien.pasdeloup-ext@st.com>
Tested-by: Sebastien PASDELOUP <sebastien.pasdeloup-ext@st.com>
---
 drivers/st/crypto/stm32_saes.c            | 223 ++++++++++------------
 include/drivers/st/stm32_cryp2_saes_reg.h | 125 ++++++++++++
 include/drivers/st/stm32_saes_reg.h       | 120 ++++++++++++
 plat/st/stm32mp2/platform.mk              |  10 +
 4 files changed, 360 insertions(+), 118 deletions(-)
 create mode 100644 include/drivers/st/stm32_cryp2_saes_reg.h
 create mode 100644 include/drivers/st/stm32_saes_reg.h

diff --git a/drivers/st/crypto/stm32_saes.c b/drivers/st/crypto/stm32_saes.c
index 23878c15a..ec8e6efa1 100644
--- a/drivers/st/crypto/stm32_saes.c
+++ b/drivers/st/crypto/stm32_saes.c
@@ -11,6 +11,11 @@
 #include <drivers/clk.h>
 #include <drivers/delay_timer.h>
 #include <drivers/st/stm32_saes.h>
+#if STM32_SAES_CRYP2
+#include <drivers/st/stm32_cryp2_saes_reg.h>
+#else /* !STM32_SAES_CRYP2 */
+#include <drivers/st/stm32_saes_reg.h>
+#endif /* STM32_SAES_CRYP2 */
 #include <drivers/st/stm32mp_reset.h>
 #include <lib/mmio.h>
 #include <lib/utils_def.h>
@@ -23,117 +28,10 @@
 #define AES_BLOCK_SIZE			(AES_BLOCK_SIZE_BIT / UINT8_BIT)
 
 #define AES_KEYSIZE_128			16U
+#define AES_KEYSIZE_192			24U
 #define AES_KEYSIZE_256			32U
 #define AES_IVSIZE			16U
 
-/* SAES control register */
-#define _SAES_CR			0x0U
-/* SAES status register */
-#define _SAES_SR			0x04U
-/* SAES data input register */
-#define _SAES_DINR			0x08U
-/* SAES data output register */
-#define _SAES_DOUTR			0x0CU
-/* SAES key registers [0-3] */
-#define _SAES_KEYR0			0x10U
-#define _SAES_KEYR1			0x14U
-#define _SAES_KEYR2			0x18U
-#define _SAES_KEYR3			0x1CU
-/* SAES initialization vector registers [0-3] */
-#define _SAES_IVR0			0x20U
-#define _SAES_IVR1			0x24U
-#define _SAES_IVR2			0x28U
-#define _SAES_IVR3			0x2CU
-/* SAES key registers [4-7] */
-#define _SAES_KEYR4			0x30U
-#define _SAES_KEYR5			0x34U
-#define _SAES_KEYR6			0x38U
-#define _SAES_KEYR7			0x3CU
-/* SAES suspend registers [0-7] */
-#define _SAES_SUSPR0			0x40U
-#define _SAES_SUSPR1			0x44U
-#define _SAES_SUSPR2			0x48U
-#define _SAES_SUSPR3			0x4CU
-#define _SAES_SUSPR4			0x50U
-#define _SAES_SUSPR5			0x54U
-#define _SAES_SUSPR6			0x58U
-#define _SAES_SUSPR7			0x5CU
-/* SAES Interrupt Enable Register */
-#define _SAES_IER			0x300U
-/* SAES Interrupt Status Register */
-#define _SAES_ISR			0x304U
-/* SAES Interrupt Clear Register */
-#define _SAES_ICR			0x308U
-
-/* SAES control register fields */
-#define _SAES_CR_RESET_VALUE		0x0U
-#define _SAES_CR_IPRST			BIT(31)
-#define _SAES_CR_KEYSEL_MASK		GENMASK(30, 28)
-#define _SAES_CR_KEYSEL_SHIFT		28U
-#define _SAES_CR_KEYSEL_SOFT		0x0U
-#define _SAES_CR_KEYSEL_DHUK		0x1U
-#define _SAES_CR_KEYSEL_BHK		0x2U
-#define _SAES_CR_KEYSEL_BHU_XOR_BH_K	0x4U
-#define _SAES_CR_KEYSEL_TEST		0x7U
-#define _SAES_CR_KSHAREID_MASK		GENMASK(27, 26)
-#define _SAES_CR_KSHAREID_SHIFT		26U
-#if STM32_SAES_VER == 0x30
-#define _SAES_CR_KSHAREID_CRYP		0x0U
-#else /* STM32_SAES_VER != 0x30 */
-#define _SAES_CR_KSHAREID_CRYP1		0x0U
-#define _SAES_CR_KSHAREID_CRYP2		0x1U
-#endif /* STM32_SAES_VER == 0x30 */
-#define _SAES_CR_KEYMOD_MASK		GENMASK(25, 24)
-#define _SAES_CR_KEYMOD_SHIFT		24U
-#define _SAES_CR_KEYMOD_NORMAL		0x0U
-#define _SAES_CR_KEYMOD_WRAPPED		0x1U
-#define _SAES_CR_KEYMOD_SHARED		0x2U
-#define _SAES_CR_NPBLB_MASK		GENMASK(23, 20)
-#define _SAES_CR_NPBLB_SHIFT		20U
-#define _SAES_CR_KEYPROT		BIT(19)
-#define _SAES_CR_KEYSIZE		BIT(18)
-#define _SAES_CR_GCMPH_MASK		GENMASK(14, 13)
-#define _SAES_CR_GCMPH_SHIFT		13U
-#define _SAES_CR_GCMPH_INIT		0U
-#define _SAES_CR_GCMPH_HEADER		1U
-#define _SAES_CR_GCMPH_PAYLOAD		2U
-#define _SAES_CR_GCMPH_FINAL		3U
-#define _SAES_CR_DMAOUTEN		BIT(12)
-#define _SAES_CR_DMAINEN		BIT(11)
-#define _SAES_CR_CHMOD_MASK		(BIT(16) | GENMASK(6, 5))
-#define _SAES_CR_CHMOD_SHIFT		5U
-#define _SAES_CR_CHMOD_ECB		0x0U
-#define _SAES_CR_CHMOD_CBC		0x1U
-#define _SAES_CR_CHMOD_CTR		0x2U
-#define _SAES_CR_CHMOD_GCM		0x3U
-#define _SAES_CR_CHMOD_GMAC		0x3U
-#define _SAES_CR_CHMOD_CCM		0x800U
-#define _SAES_CR_MODE_MASK		GENMASK(4, 3)
-#define _SAES_CR_MODE_SHIFT		3U
-#define _SAES_CR_MODE_ENC		0U
-#define _SAES_CR_MODE_KEYPREP		1U
-#define _SAES_CR_MODE_DEC		2U
-#define _SAES_CR_DATATYPE_MASK		GENMASK(2, 1)
-#define _SAES_CR_DATATYPE_SHIFT		1U
-#define _SAES_CR_DATATYPE_NONE		0U
-#define _SAES_CR_DATATYPE_HALF_WORD	1U
-#define _SAES_CR_DATATYPE_BYTE		2U
-#define _SAES_CR_DATATYPE_BIT		3U
-#define _SAES_CR_EN			BIT(0)
-
-/* SAES status register fields */
-#define _SAES_SR_KEYVALID		BIT(7)
-#define _SAES_SR_BUSY			BIT(3)
-#define _SAES_SR_WRERR			BIT(2)
-#define _SAES_SR_RDERR			BIT(1)
-#define _SAES_SR_CCF			BIT(0)
-
-/* SAES interrupt registers fields */
-#define _SAES_I_RNG_ERR			BIT(3)
-#define _SAES_I_KEY_ERR			BIT(2)
-#define _SAES_I_RW_ERR			BIT(1)
-#define _SAES_I_CC			BIT(0)
-
 #define SAES_TIMEOUT_US			100000U
 #define TIMEOUT_US_1MS			1000U
 #define SAES_RESET_DELAY		20U
@@ -196,7 +94,7 @@ static int wait_computation_completed(uintptr_t base)
 {
 	uint64_t timeout = timeout_init_us(SAES_TIMEOUT_US);
 
-	while ((mmio_read_32(base + _SAES_SR) & _SAES_SR_CCF) != _SAES_SR_CCF) {
+	while ((mmio_read_32(base + _SAES_ISR) & _SAES_I_CCF) != _SAES_I_CCF) {
 		if (timeout_elapsed(timeout)) {
 			WARN("%s: timeout\n", __func__);
 			return -ETIMEDOUT;
@@ -208,7 +106,21 @@ static int wait_computation_completed(uintptr_t base)
 
 static void clear_computation_completed(uintptr_t base)
 {
-	mmio_setbits_32(base + _SAES_ICR, _SAES_I_CC);
+	mmio_setbits_32(base + _SAES_ICR, _SAES_I_CCF);
+}
+
+static int wait_key_valid(uintptr_t base)
+{
+	uint64_t timeout = timeout_init_us(SAES_TIMEOUT_US);
+
+	while ((mmio_read_32(base + _SAES_SR) & _SAES_SR_KEYVALID) != _SAES_SR_KEYVALID) {
+		if (timeout_elapsed(timeout)) {
+			WARN("%s: timeout\n", __func__);
+			return -ETIMEDOUT;
+		}
+	}
+
+	return 0;
 }
 
 static int saes_start(struct stm32_saes_context *ctx)
@@ -258,20 +170,58 @@ static void saes_write_iv(struct stm32_saes_context *ctx)
 
 }
 
+static uint32_t saes_get_keysize(struct stm32_saes_context *ctx)
+{
+	int keysize;
+	uint32_t ret = 0U;
+
+#if STM32_SAES_CRYP2
+	keysize = (ctx->cr & _SAES_CR_KEYSIZE_MASK) >> _SAES_CR_KEYSIZE_SHIFT;
+#else /* !STM32_SAES_CRYP2 */
+	keysize = (ctx->cr & _SAES_CR_KEYSIZE) >> _SAES_CR_KEYSIZE_SHIFT;
+#endif /* STM32_SAES_CRYP2 */
+
+	switch (keysize) {
+	case _SAES_CR_KEYSIZE_256:
+		ret = AES_KEYSIZE_256;
+		break;
+#if STM32_SAES_CRYP2
+	case _SAES_CR_KEYSIZE_192:
+		ret = AES_KEYSIZE_192;
+		break;
+#endif /* STM32_SAES_CRYP2 */
+	case _SAES_CR_KEYSIZE_128:
+		ret = AES_KEYSIZE_128;
+		break;
+	default:
+		ERROR("Invalid keysize\n");
+		break;
+	}
+
+	return ret;
+}
+
 static void saes_write_key(struct stm32_saes_context *ctx)
 {
 	/* Restore the _SAES_KEYRx if SOFTWARE key */
 	if ((ctx->cr & _SAES_CR_KEYSEL_MASK) == (_SAES_CR_KEYSEL_SOFT << _SAES_CR_KEYSEL_SHIFT)) {
 		uint8_t i;
+		uint32_t ks;
+
+		ks = saes_get_keysize(ctx);
+		if (ks == 0U) {
+			panic();
+		}
 
-		for (i = 0U; i < AES_KEYSIZE_128 / sizeof(uint32_t); i++) {
+		for (i = 0; i < (AES_KEYSIZE_128 / sizeof(uint32_t)); i++) {
 			mmio_write_32(ctx->base + _SAES_KEYR0 + i * sizeof(uint32_t), ctx->key[i]);
 		}
 
-		if ((ctx->cr & _SAES_CR_KEYSIZE) == _SAES_CR_KEYSIZE) {
-			for (i = 0U; i < (AES_KEYSIZE_256 / 2U) / sizeof(uint32_t); i++) {
+		if (ks > AES_KEYSIZE_128) {
+			ks -= AES_KEYSIZE_128; /* writes only remaining bytes */
+			for (i = 0; i < (ks / sizeof(uint32_t)); i++) {
 				mmio_write_32(ctx->base + _SAES_KEYR4 + i * sizeof(uint32_t),
-					      ctx->key[i + 4U]);
+					      ctx->key[i + 4]);
 			}
 		}
 	}
@@ -279,22 +229,30 @@ static void saes_write_key(struct stm32_saes_context *ctx)
 
 static int saes_prepare_key(struct stm32_saes_context *ctx)
 {
+	int ret;
+
 	/* Disable the SAES peripheral */
 	mmio_clrbits_32(ctx->base + _SAES_CR, _SAES_CR_EN);
 
+#if !STM32_SAES_CRYP2
 	/* Set key size */
 	if ((ctx->cr & _SAES_CR_KEYSIZE) != 0U) {
 		mmio_setbits_32(ctx->base + _SAES_CR, _SAES_CR_KEYSIZE);
 	} else {
 		mmio_clrbits_32(ctx->base + _SAES_CR, _SAES_CR_KEYSIZE);
 	}
+#endif /* !STM32_SAES_CRYP2 */
 
 	saes_write_key(ctx);
 
+	ret = wait_key_valid(ctx->base);
+	if (ret != 0) {
+		return ret;
+	}
+
 	/* For ECB/CBC decryption, key preparation mode must be selected to populate the key */
 	if ((IS_CHAINING_MODE(ECB, ctx->cr) || IS_CHAINING_MODE(CBC, ctx->cr)) &&
 	    is_decrypt(ctx->cr)) {
-		int ret;
 
 		/* Select Mode 2 */
 		mmio_clrsetbits_32(ctx->base + _SAES_CR, _SAES_CR_MODE_MASK,
@@ -321,7 +279,7 @@ static int saes_prepare_key(struct stm32_saes_context *ctx)
 
 static int save_context(struct stm32_saes_context *ctx)
 {
-	if ((mmio_read_32(ctx->base + _SAES_SR) & _SAES_SR_CCF) != 0U) {
+	if ((mmio_read_32(ctx->base + _SAES_ISR) & _SAES_I_CCF) != 0U) {
 		/* Device should not be in a processing phase */
 		return -EINVAL;
 	}
@@ -385,7 +343,7 @@ int stm32_saes_driver_init(void)
 {
 	int err;
 
-	if(saes_pdata.base != 0U) {
+	if (saes_pdata.base != 0U) {
 		/* Driver is already initialized */
 		return 0;
 	}
@@ -486,6 +444,24 @@ int stm32_saes_init(struct stm32_saes_context *ctx, bool is_dec,
 			   _SAES_CR_DATATYPE_BYTE << _SAES_CR_DATATYPE_SHIFT);
 
 	/* Configure keysize */
+#if STM32_SAES_CRYP2
+	switch (key_size) {
+	case AES_KEYSIZE_128:
+		mmio_clrsetbits_32((uintptr_t)&(ctx->cr), _SAES_CR_KEYSIZE_MASK,
+				   _SAES_CR_KEYSIZE_128 << _SAES_CR_KEYSIZE_SHIFT);
+		break;
+	case AES_KEYSIZE_192:
+		mmio_clrsetbits_32((uintptr_t)&(ctx->cr), _SAES_CR_KEYSIZE_MASK,
+				   _SAES_CR_KEYSIZE_192 << _SAES_CR_KEYSIZE_SHIFT);
+		break;
+	case AES_KEYSIZE_256:
+		mmio_clrsetbits_32((uintptr_t)&(ctx->cr), _SAES_CR_KEYSIZE_MASK,
+				   _SAES_CR_KEYSIZE_256 << _SAES_CR_KEYSIZE_SHIFT);
+		break;
+	default:
+		return -EINVAL;
+	}
+#else /* !STM32_SAES_CRYP2 */
 	switch (key_size) {
 	case AES_KEYSIZE_128:
 		mmio_clrbits_32((uintptr_t)&(ctx->cr), _SAES_CR_KEYSIZE);
@@ -496,6 +472,7 @@ int stm32_saes_init(struct stm32_saes_context *ctx, bool is_dec,
 	default:
 		return -EINVAL;
 	}
+#endif /* STM32_SAES_CRYP2 */
 
 	/* Configure key */
 	switch (key_select) {
@@ -506,13 +483,23 @@ int stm32_saes_init(struct stm32_saes_context *ctx, bool is_dec,
 		switch (key_size) {
 		case AES_KEYSIZE_128:
 			/* First 16 bytes == 4 u32 */
-			for (i = 0U; i < AES_KEYSIZE_128 / sizeof(uint32_t); i++) {
+			for (i = 0U; i < (AES_KEYSIZE_128 / sizeof(uint32_t)); i++) {
 				mmio_write_32((uintptr_t)(ctx->key + i), htobe32(key_u32[3 - i]));
 				/* /!\ we save the key in HW byte order
 				 * and word order : key[i] is for _SAES_KEYRi
 				 */
 			}
 			break;
+#if STM32_SAES_CRYP2
+		case AES_KEYSIZE_192:
+			for (i = 0U; i < AES_KEYSIZE_192 / sizeof(uint32_t); i++) {
+				mmio_write_32((uintptr_t)(ctx->key + i), htobe32(key_u32[5 - i]));
+				/* /!\ we save the key in HW byte order
+				 * and word order : key[i] is for _SAES_KEYRi
+				 */
+			}
+			break;
+#endif /* STM32_SAES_CRYP2 */
 		case AES_KEYSIZE_256:
 			for (i = 0U; i < AES_KEYSIZE_256 / sizeof(uint32_t); i++) {
 				mmio_write_32((uintptr_t)(ctx->key + i), htobe32(key_u32[7 - i]));
diff --git a/include/drivers/st/stm32_cryp2_saes_reg.h b/include/drivers/st/stm32_cryp2_saes_reg.h
new file mode 100644
index 000000000..0c7eaf74b
--- /dev/null
+++ b/include/drivers/st/stm32_cryp2_saes_reg.h
@@ -0,0 +1,125 @@
+/*
+ * Copyright (c) 2024, STMicroelectronics - All Rights Reserved
+ *
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+#ifndef STM32_CRYP2_SAES_REG_H
+#define STM32_CRYP2_SAES_REG_H
+
+/* SAES control register */
+#define _SAES_CR			0x0U
+/* SAES status register */
+#define _SAES_SR			0x04U
+/* SAES data input register */
+#define _SAES_DINR			0x08U
+/* SAES data output register */
+#define _SAES_DOUTR			0x0CU
+/* SAES key registers [0-7] */
+#define _SAES_KEYR0			0x20U
+#define _SAES_KEYR1			0x24U
+#define _SAES_KEYR2			0x28U
+#define _SAES_KEYR3			0x2CU
+#define _SAES_KEYR4			0x30U
+#define _SAES_KEYR5			0x34U
+#define _SAES_KEYR6			0x38U
+#define _SAES_KEYR7			0x3CU
+/* SAES initialization vector registers [0-3] */
+#define _SAES_IVR0			0x40U
+#define _SAES_IVR1			0x44U
+#define _SAES_IVR2			0x48U
+#define _SAES_IVR3			0x4CU
+/* SAES context swap registers [0-7] */
+#define _SAES_CSR0			0x50U
+#define _SAES_CSR1			0x54U
+#define _SAES_CSR2			0x58U
+#define _SAES_CSR3			0x5CU
+#define _SAES_CSR4			0x50U
+#define _SAES_CSR5			0x54U
+#define _SAES_CSR6			0x58U
+#define _SAES_CSR7			0x5CU
+/* SAES Interrupt Enable Register */
+#define _SAES_IER			0x300U
+/* SAES Interrupt Status Register */
+#define _SAES_ISR			0x304U
+/* SAES Interrupt Clear Register */
+#define _SAES_ICR			0x308U
+
+/* SAES control register fields */
+#define _SAES_CR_RESET_VALUE		0x0U
+#define _SAES_CR_IPRST			BIT(31)
+#define _SAES_CR_KEYSEL_MASK		GENMASK_32(30, 28)
+#define _SAES_CR_KEYSEL_SHIFT		28U
+#define _SAES_CR_KEYSEL_SOFT		0x0U
+#define _SAES_CR_KEYSEL_DHUK		0x1U
+#define _SAES_CR_KEYSEL_BHK		0x2U
+#define _SAES_CR_KEYSEL_BHU_XOR_BH_K	0x4U
+#define _SAES_CR_WRAPID_MASK		GENMASK_32(27, 25)
+#define _SAES_CR_WRAPID_SHIFT		25U
+#define _SAES_CR_WRAPID_SAES		0x0U
+#define _SAES_CR_WRAPID_CRYP1		0x1U
+#define _SAES_CR_WRAPID_CRYP2		0x2U
+#define _SAES_CR_WRAPEN			BIT(24)
+#define _SAES_CR_NPBLB_MASK		GENMASK_32(23, 20)
+#define _SAES_CR_NPBLB_SHIFT		20U
+#define _SAES_CR_KEYPROT		BIT(19)
+#define _SAES_CR_KEYSIZE_MASK		GENMASK_32(18, 17)
+#define _SAES_CR_KEYSIZE_SHIFT		17U
+#define _SAES_CR_KEYSIZE_128		0x0U
+#define _SAES_CR_KEYSIZE_192		0x1U
+#define _SAES_CR_KEYSIZE_256		0x2U
+#define _SAES_CR_CPHASE_MASK		GENMASK_32(14, 13)
+#define _SAES_CR_CPHASE_SHIFT		13U
+#define _SAES_CR_CPHASE_INIT		0U
+#define _SAES_CR_CPHASE_HEADER		1U
+#define _SAES_CR_CPHASE_PAYLOAD		2U
+#define _SAES_CR_CPHASE_FINAL		3U
+#define _SAES_CR_DMAOUTEN		BIT(12)
+#define _SAES_CR_DMAINEN		BIT(11)
+#define _SAES_CR_CHMOD_MASK		GENMASK_32(8, 5)
+#define _SAES_CR_CHMOD_SHIFT		5U
+#define _SAES_CR_CHMOD_ECB		0x0U
+#define _SAES_CR_CHMOD_CBC		0x1U
+#define _SAES_CR_CHMOD_CTR		0x2U
+#define _SAES_CR_CHMOD_GCM		0x3U
+#define _SAES_CR_CHMOD_GMAC		0x3U
+#define _SAES_CR_CHMOD_CCM		0x4U
+#define _SAES_CR_OPMODE_MASK		GENMASK_32(4, 3)
+#define _SAES_CR_OPMODE_SHIFT		3U
+#define _SAES_CR_OPMODE_ENC		0U
+#define _SAES_CR_OPMODE_KEYPREP		1U
+#define _SAES_CR_OPMODE_DEC		2U
+#define _SAES_CR_DATATYPE_MASK		GENMASK_32(2, 1)
+#define _SAES_CR_DATATYPE_SHIFT		1U
+#define _SAES_CR_DATATYPE_NONE		0U
+#define _SAES_CR_DATATYPE_HALF_WORD	1U
+#define _SAES_CR_DATATYPE_BYTE		2U
+#define _SAES_CR_DATATYPE_BIT		3U
+#define _SAES_CR_EN			BIT(0)
+
+/* SAES status register fields */
+#define _SAES_SR_KEYVALID		BIT(7)
+#define _SAES_SR_WRERR			BIT(6)
+#define _SAES_SR_RDERR			BIT(5)
+#define _SAES_SR_BUSY			BIT(4)
+
+/* SAES interrupt registers fields */
+#define _SAES_I_RNG_ERR			BIT(5)
+#define _SAES_I_KEY_ERR			BIT(4)
+#define _SAES_I_RW_ERR			BIT(3)
+#define _SAES_I_CCF			BIT(2)
+
+/* Correspondence between registers fields and the used define in the driver */
+#define _SAES_CR_GCMPH_MASK		_SAES_CR_CPHASE_MASK
+#define _SAES_CR_GCMPH_SHIFT		_SAES_CR_CPHASE_SHIFT
+#define _SAES_CR_GCMPH_INIT		_SAES_CR_CPHASE_INIT
+#define _SAES_CR_GCMPH_HEADER		_SAES_CR_CPHASE_HEADER
+#define _SAES_CR_GCMPH_PAYLOAD		_SAES_CR_CPHASE_PAYLOAD
+#define _SAES_CR_GCMPH_FINAL		_SAES_CR_CPHASE_FINAL
+
+#define _SAES_CR_MODE_MASK		_SAES_CR_OPMODE_MASK
+#define _SAES_CR_MODE_SHIFT		_SAES_CR_OPMODE_SHIFT
+#define _SAES_CR_MODE_ENC		_SAES_CR_OPMODE_ENC
+#define _SAES_CR_MODE_KEYPREP		_SAES_CR_OPMODE_KEYPREP
+#define _SAES_CR_MODE_DEC		_SAES_CR_OPMODE_DEC
+
+#endif /* STM32_CRYP2_SAES_REG_H */
\ No newline at end of file
diff --git a/include/drivers/st/stm32_saes_reg.h b/include/drivers/st/stm32_saes_reg.h
new file mode 100644
index 000000000..6cf5a16d9
--- /dev/null
+++ b/include/drivers/st/stm32_saes_reg.h
@@ -0,0 +1,120 @@
+/*
+ * Copyright (c) 2024, STMicroelectronics - All Rights Reserved
+ *
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+#ifndef STM32_SAES_REG_H
+#define STM32_SAES_REG_H
+
+/* SAES control register */
+#define _SAES_CR			0x0U
+/* SAES status register */
+#define _SAES_SR			0x04U
+/* SAES data input register */
+#define _SAES_DINR			0x08U
+/* SAES data output register */
+#define _SAES_DOUTR			0x0CU
+/* SAES key registers [0-3] */
+#define _SAES_KEYR0			0x10U
+#define _SAES_KEYR1			0x14U
+#define _SAES_KEYR2			0x18U
+#define _SAES_KEYR3			0x1CU
+/* SAES initialization vector registers [0-3] */
+#define _SAES_IVR0			0x20U
+#define _SAES_IVR1			0x24U
+#define _SAES_IVR2			0x28U
+#define _SAES_IVR3			0x2CU
+/* SAES key registers [4-7] */
+#define _SAES_KEYR4			0x30U
+#define _SAES_KEYR5			0x34U
+#define _SAES_KEYR6			0x38U
+#define _SAES_KEYR7			0x3CU
+/* SAES suspend registers [0-7] */
+#define _SAES_SUSPR0			0x40U
+#define _SAES_SUSPR1			0x44U
+#define _SAES_SUSPR2			0x48U
+#define _SAES_SUSPR3			0x4CU
+#define _SAES_SUSPR4			0x50U
+#define _SAES_SUSPR5			0x54U
+#define _SAES_SUSPR6			0x58U
+#define _SAES_SUSPR7			0x5CU
+/* SAES Interrupt Enable Register */
+#define _SAES_IER			0x300U
+/* SAES Interrupt Status Register */
+#define _SAES_ISR			0x304U
+/* SAES Interrupt Clear Register */
+#define _SAES_ICR			0x308U
+
+/* SAES control register fields */
+#define _SAES_CR_RESET_VALUE		0x0U
+#define _SAES_CR_IPRST			BIT(31)
+#define _SAES_CR_KEYSEL_MASK		GENMASK_32(30, 28)
+#define _SAES_CR_KEYSEL_SHIFT		28U
+#define _SAES_CR_KEYSEL_SOFT		0x0U
+#define _SAES_CR_KEYSEL_DHUK		0x1U
+#define _SAES_CR_KEYSEL_BHK		0x2U
+#define _SAES_CR_KEYSEL_BHU_XOR_BH_K	0x4U
+#define _SAES_CR_KEYSEL_TEST		0x7U
+#define _SAES_CR_KSHAREID_MASK		GENMASK_32(27, 26)
+#define _SAES_CR_KSHAREID_SHIFT		26U
+#if STM32_SAES_VER == 0x30
+#define _SAES_CR_KSHAREID_CRYP		0x0U
+#else /* STM32_SAES_VER == 0x30 */
+#define _SAES_CR_KSHAREID_CRYP1		0x0U
+#define _SAES_CR_KSHAREID_CRYP2		0x1U
+#endif /* STM32_SAES_VER == 0x30 */
+#define _SAES_CR_KEYMOD_MASK		GENMASK_32(25, 24)
+#define _SAES_CR_KEYMOD_SHIFT		24U
+#define _SAES_CR_KEYMOD_NORMAL		0x0U
+#define _SAES_CR_KEYMOD_WRAPPED		0x1U
+#define _SAES_CR_KEYMOD_SHARED		0x2U
+#define _SAES_CR_NPBLB_MASK		GENMASK_32(23, 20)
+#define _SAES_CR_NPBLB_SHIFT		20U
+#define _SAES_CR_KEYPROT		BIT(19)
+#define _SAES_CR_KEYSIZE		BIT(18)
+#define _SAES_CR_KEYSIZE_SHIFT          18
+#define _SAES_CR_KEYSIZE_128		0x0U
+#define _SAES_CR_KEYSIZE_256		0x1U
+#define _SAES_CR_GCMPH_MASK		GENMASK_32(14, 13)
+#define _SAES_CR_GCMPH_SHIFT		13U
+#define _SAES_CR_GCMPH_INIT		0U
+#define _SAES_CR_GCMPH_HEADER		1U
+#define _SAES_CR_GCMPH_PAYLOAD		2U
+#define _SAES_CR_GCMPH_FINAL		3U
+#define _SAES_CR_DMAOUTEN		BIT(12)
+#define _SAES_CR_DMAINEN		BIT(11)
+#define _SAES_CR_CHMOD_MASK		(BIT(16) | GENMASK_32(6, 5))
+#define _SAES_CR_CHMOD_SHIFT		5U
+#define _SAES_CR_CHMOD_ECB		0x0U
+#define _SAES_CR_CHMOD_CBC		0x1U
+#define _SAES_CR_CHMOD_CTR		0x2U
+#define _SAES_CR_CHMOD_GCM		0x3U
+#define _SAES_CR_CHMOD_GMAC		0x3U
+#define _SAES_CR_CHMOD_CCM		0x800U
+#define _SAES_CR_MODE_MASK		GENMASK_32(4, 3)
+#define _SAES_CR_MODE_SHIFT		3U
+#define _SAES_CR_MODE_ENC		0U
+#define _SAES_CR_MODE_KEYPREP		1U
+#define _SAES_CR_MODE_DEC		2U
+#define _SAES_CR_DATATYPE_MASK		GENMASK_32(2, 1)
+#define _SAES_CR_DATATYPE_SHIFT		1U
+#define _SAES_CR_DATATYPE_NONE		0U
+#define _SAES_CR_DATATYPE_HALF_WORD	1U
+#define _SAES_CR_DATATYPE_BYTE		2U
+#define _SAES_CR_DATATYPE_BIT		3U
+#define _SAES_CR_EN			BIT(0)
+
+/* SAES status register fields */
+#define _SAES_SR_KEYVALID		BIT(7)
+#define _SAES_SR_BUSY			BIT(3)
+#define _SAES_SR_WRERR			BIT(2)
+#define _SAES_SR_RDERR			BIT(1)
+#define _SAES_SR_CCF			BIT(0)
+
+/* SAES interrupt registers fields */
+#define _SAES_I_RNG_ERR			BIT(3)
+#define _SAES_I_KEY_ERR			BIT(2)
+#define _SAES_I_RW_ERR			BIT(1)
+#define _SAES_I_CCF			BIT(0)
+
+#endif /* STM32_SAES_REG_H */
\ No newline at end of file
diff --git a/plat/st/stm32mp2/platform.mk b/plat/st/stm32mp2/platform.mk
index 7c25958e7..f76f73b38 100644
--- a/plat/st/stm32mp2/platform.mk
+++ b/plat/st/stm32mp2/platform.mk
@@ -84,6 +84,14 @@ STM32_HASH_VER			:=	4
 STM32_RNG_VER			:=	4
 STM32_SAES_VER			:=	50 # 0x32
 
+ifeq ($(STM32MP21),1)
+STM32_SAES_VER			:=	16 # 0x10
+STM32_SAES_CRYP2		:=	1
+else
+STM32_SAES_VER			:=	50 # 0x32
+STM32_SAES_CRYP2		:=	0
+endif
+
 # Set load address for serial boot devices
 DWL_BUFFER_BASE 		?=	0x87000000
 
@@ -168,6 +176,7 @@ $(eval $(call assert_booleans,\
 		PKA_USE_BRAINPOOL_P256T1 \
 		PKA_USE_NIST_P256 \
 		STM32MP_CRYPTO_ROM_LIB \
+		STM32_SAES_CRYP2 \
 		STM32MP_DDR_DUAL_AXI_PORT \
 		STM32MP_DDR_FIP_IO_STORAGE \
 		STM32MP_DDR3_TYPE \
@@ -200,6 +209,7 @@ $(eval $(call add_defines,\
 		STM32_HASH_VER \
 		STM32_RNG_VER \
 		STM32_SAES_VER \
+		STM32_SAES_CRYP2 \
 		STM32_TF_A_COPIES \
 		STM32MP_CRYPTO_ROM_LIB \
 		STM32MP_DDR_DUAL_AXI_PORT \
-- 
2.47.2

